Display list of requesters in Library

Taranjot-Singh opened the pull request:
Major Things to Notice in this PR:

1. The Library Repository now has a LiveData<List<User>>  to keep track of requesters on a book that will be displayed in the detailed book view. This also means the addition of another listernerRegistration that will listen to changes in the database and update the list of requesters if requests on the book are added/removed.

2. Unlike list of books in the Library, where the snapshot listener is attached to the same query throughout the time that the app is running, the list of requesters changes depending on which book is selected by the user and so the snapshot listener must change the query it listens to. For this reason, it makes sense to make a query and to attach the listener for a specific book when we enter the detailed book view fragment for that book. And when we exit out of the detailed book view fragment, we no longer need to listen to changes in the requesters list for that book and we detach the listener. Thus, in the LibraryBookDetails Fragment, you will find the call to functions that do this in the onCreateView() and onDestroyView() methods.

3. The Library ViewModel has some minor functions added that facilitate passing function calls from the LibraryDetailedBookFragment to the repository and facilitate passing the list of requesters from the repository to the LibraryDetailedBookFragment

4. The xml file for how a requester's info should appear has been added. An adapter class for the recycler view of requesters has also been added.

5. The recycler view list of requesters was added below all the other views that alredy existed in the constraint layout in the LibraryDetailedBookView,. In order to make the whole fragment scrollable as one entire page (compared to only scrolling the list of requesters at the bottom of the page), the constraint layout has now been enclosed inside <ScrollView> tags. Because of this the height of the ImageView in the LibraryDetailedBookView had to be changed from a percentage to a hardcoded number. If we use a percentage, the height of the image would change when the number of items in the requester list change since it is a percentage of the whole entire "page".


Here is a screenshot. The list of requesters will change dynamically if another request is added for that book even while the owner is viewing the list of requesters:
![image](https://user-images.githubusercontent.com/43302693/98626375-7400a380-22c6-11eb-903b-a206ce83f318.png)


# TODO
- Add further functionality to display a fragment with requester's profile when user clicks one in the list. This will  involve navigating to a new fragment and likely involve adding a LiveData<User> and another snapshot listener to Library Repository so that we can update the details shown in the requester's profile segment if it changes in the database. It will also have the buttons to accept/decline a request.
- Currently, the image used for requesters is just a placeholder until we start storing user's images in firestore (are we going to so this?)

calebschoepp commented on app/src/main/java/com/example/unlibrary/library/LibraryBookDetailsFragment.java:
```diff
@@ -43,7 +46,7 @@
      */
     @Override
     public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
-        // Get the activity viewModel
+
```
Removing comment is fine but Nit: no line before first line of function.

calebschoepp commented on app/src/main/java/com/example/unlibrary/library/LibraryBookDetailsFragment.java:
```diff
@@ -65,6 +68,25 @@ public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                     .show();
         });
 
+        //Setup the list of requesters
```
Mega super nit: IMO I prefer a space before the text of a comment.

```suggestion
        // Setup the list of requesters
```

calebschoepp commented on app/src/main/java/com/example/unlibrary/library/LibraryBookDetailsFragment.java:
```diff
@@ -65,6 +68,25 @@ public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
+    @Override
+    public void onDestroyView() {
+        super.onDestroyView();
+        LibraryViewModel mViewModel = new ViewModelProvider(requireActivity()).get(LibraryViewModel.class);
```
`mViewModel` should be a member variable of the class so that you don't need to get the viewModel twice.

calebschoepp commented on app/src/main/java/com/example/unlibrary/library/LibraryRepository.java:
```diff
@@ -179,4 +188,72 @@ public void setFilter(FilterMap filter) {
+     * @param currentBookID
+     */
+    public void fetchRequestersForCurrentBook(String currentBookID) {
+        mCurrentBookRequesters.setValue(new ArrayList<>()); //Is this required to ensure we clear the previous book's requesters in time before requesters get displayed?
```
Probably? IMO can't hurt to leave it.

calebschoepp commented on app/src/main/java/com/example/unlibrary/library/RequestersRecyclerViewAdapter.java:
```diff
@@ -0,0 +1,110 @@
+package com.example.unlibrary.library;
```
File header please.

calebschoepp commented on app/src/main/res/layout/fragment_library_requester_item.xml:
```diff
@@ -0,0 +1,62 @@
+            android:textAppearance="@style/TextAppearance.MaterialComponents.Body2"
+            app:layout_constraintStart_toEndOf="@id/userImage"
+            app:layout_constraintTop_toBottomOf="@+id/requesterUsername"
+            tools:text="email@email.com Ici" />
```
What is this?

calebschoepp reviewed (approved):
Couple of comments but in general I think it looks good. Also, I don't think we are planning on adding profile photos for users.

LGTM

Taranjot-Singh commented:
Minor fixes:
- Removed the user_icon next to each item requesters list and instead added divider between list items
- Addressed Caleb's comments
- Fixed comments

TODO:
Remove user icon from the exchange fragment as well that was merged into master by golnoush's PR

cyrusdiego reviewed (approved):
LGTM! Good work on this :D
